python/mpt/src/mpt/hash_diff.cairo

```diff
@@ -47,3 +59,97 @@ func poseidon_storage_diff{poseidon_ptr: PoseidonBuiltin*}(diff: StorageDiffEntr
     let (storage_diff_hash) = poseidon_hash_many(5, buffer);
     return storage_diff_hash;
 }
+
+func hash_account_diff_segment{poseidon_ptr: PoseidonBuiltin*}(account_diff: AccountDiff) -> felt {
+    alloc_locals;
+    let len = account_diff.value.len;
+    if (len == 0) {
+        // TODO: Do we return 0 as the hash of an empty segment?
+        // We don't want to raise here as a block could have 0 account diffs.
+        return 0;
+    }
+    let accumulator = poseidon_account_diff(account_diff.value.data[0]);
+    let final_hash = _hash_account_diff_inner(accumulator, account_diff.value.data, 1, len);
+    return final_hash;
+}
+
+func _hash_account_diff_inner{poseidon_ptr: PoseidonBuiltin*}(
+    accumulator: felt, start_ptr: AddressAccountDiffEntry*, i: felt, len: felt
+) -> felt {
+    if (i == len) {
+        return accumulator;
+    }
+    let next_hash = poseidon_account_diff(start_ptr[i]);
+
+    let (buffer) = alloc();
+    assert buffer[0] = accumulator;
+    assert buffer[1] = next_hash;
+    let (accumulator) = poseidon_hash_many(2, buffer);
+
+    return _hash_account_diff_inner(accumulator, start_ptr, i + 1, len);
+}
+
+func hash_state_account_diff{poseidon_ptr: PoseidonBuiltin*}(state: State) -> felt {
+    alloc_locals;
+    let dict_ptr_start = state.value._main_trie.value._data.value.dict_ptr_start;
+    let dict_ptr_end = state.value._main_trie.value._data.value.dict_ptr;
+    let len = dict_ptr_end - dict_ptr_start;
+    if (len == 0) {
+        return 0;
+    }
+
+    // We cast the state dict pointer to an AddressAccountDiffEntry pointer as the two underlying types are identical.
+    let casted_dict_ptr_start = cast(dict_ptr_start, AddressAccountDiffEntryStruct*);
+    let accumulator = poseidon_account_diff(AddressAccountDiffEntry(casted_dict_ptr_start));
+    tempvar start_ptr = new AddressAccountDiffEntry(casted_dict_ptr_start);
+    let final_hash = _hash_account_diff_inner(accumulator, start_ptr, 1, len);
+    return final_hash;
+}
+
+func hash_storage_diff_segment{poseidon_ptr: PoseidonBuiltin*}(storage_diff: StorageDiff) -> felt {
+    alloc_locals;
+    let len = storage_diff.value.len;
+    if (len == 0) {
+        return 0;
+    }
+    let accumulator = poseidon_storage_diff(storage_diff.value.data[0]);
+    let final_hash = _hash_storage_diff_inner(accumulator, storage_diff.value.data, 1, len);
+    return final_hash;
+}
+
+func _hash_storage_diff_inner{poseidon_ptr: PoseidonBuiltin*}(
+    accumulator: felt, start_ptr: StorageDiffEntry*, i: felt, len: felt
+) -> felt {
+    if (i == len) {
+        return accumulator;
+    }
+    let next_hash = poseidon_storage_diff(start_ptr[i]);
+
+    let (buffer) = alloc();
+    assert buffer[0] = accumulator;
+    assert buffer[1] = next_hash;
+    let (accumulator) = poseidon_hash_many(2, buffer);
+
+    return _hash_storage_diff_inner(accumulator, start_ptr, i + 1, len);
+}
```

Comment:
can we make this generic by passing a `codeoffset` and doing `call abs hash_util` instead?